File: minimumTotal.py
# Time Complexity : O(mn)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        if len(triangle) < 2:
            return triangle[0][0]

        for i in range(len(triangle)-1, 1, -1):
            for k in range(len(triangle[i])-1):
                triangle[i-1][k] += min(triangle[i][k], triangle[i][k+1])

        return triangle[0][0] + min(triangle[1][0], triangle[1][1])

File: minimumTotal.java
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        if(triangle == null || triangle.size() == 0) return 0;
        int [] dp = new int [triangle.size()];
        int n = triangle.size();
        //fill the last row 
        for(int i = 0; i < dp.length; i++){
            dp[i] = triangle.get(n - 1).get(i);
        }
        
        //start from the last second row 
        for(int i = dp.length - 2; i >= 0; i--){
            for(int j = 0; j <= i; j++){
                dp[j] = Math.min(dp[j], dp[j + 1]) + triangle.get(i).get(j);
            }
        }
        return dp[0];
    }
}
